addons/sale/tests/test_sale_order.py

```diff
@@ -772,6 +772,40 @@ def test_price_recomputation_on_readonly_unit_price(self):
         })
         self.assertEqual(new_order.order_line.price_unit, 22.0)
 
+    def test_sale_order_unit_price_recompute_on_product_change(self):
+        """Ensure price_unit is correctly recomputed when the product is
+           changed after manually changing the price.
+        """
+        product2, product3 = self.env['product.product'].create([
+            {
+                'name': "Test Product2",
+                'list_price': 50.0,
+            },
+            {
+                'name': "Test Product3",
+                'list_price': 0.0,
+            }
+        ])
+        # Manually change the price on the original SO line
+        with Form(self.sale_order) as so_form:
+            with so_form.order_line.edit(0) as sol:
+                sol.price_unit = 100
+                sol.product_id = product2
+
+                # Expected price_total = list_price * quantity + tax
+                expected_total = (product2.list_price * sol.product_uom_qty) + sol.price_tax
+                self.assertAlmostEqual(
+                    sol.price_total, expected_total,
+                    msg="price_total should be equal to expected_total",
+                )
+
+                sol.product_id = product3
+                expected_total = (product3.list_price * sol.product_uom_qty) + sol.price_tax
+                self.assertAlmostEqual(
+                    sol.price_total, expected_total,
+                    msg="price_total should be equal to expected_total",
+                )
```

Comment:
Modifying the test to set a custom price on a zero-priced product, it fails with the current change:
```suggestion
        product2 = self.env['product.product'].create({
            'name': "Test Product2",
            'list_price': 0.0,
        })
        sol = self.sale_order.order_line[0]
        # Manually change the price on the original SO line
        with Form(sol) as sol_form:
            sol_form.product_id = product2
            sol_form.price_unit = 100

        # Expected price_subtotal = custom unit price * quantity
        self.assertAlmostEqual(
            sol.price_subtotal, 100 * sol.product_uom_qty,
            msg="price_total should be equal to expected_total",
        )

        with Form(sol) as sol_form:
            sol_form.product_id = self.product
        # Expected price_subtotal = list price * quantity
        self.assertAlmostEqual(
            sol.price_subtotal, self.product.list_price * sol.product_uom_qty,
            msg="price_total should be equal to expected_total",
        )
```